File: tests/plates/gopal2010/gopal2010.m
% MBDyn (C) is a multibody analysis code. 
% http://www.mbdyn.org
% 
% Copyright (C) 1996-2017
% 
% Pierangelo Masarati	<masarati@aero.polimi.it>
% Paolo Mantegazza	<mantegazza@aero.polimi.it>
% 
% Dipartimento di Ingegneria Aerospaziale - Politecnico di Milano
% via La Masa, 34 - 20156 Milano, Italy
% http://www.aero.polimi.it
% 
% Changing this copyright notice is forbidden.
% 
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation (version 2 of the License).
% 
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
% 
% Based on Pradeep Gopalakrishnan and Danesh K. Tafti
% "Effect of wing flexibility on lift and thrust production in flapping flight."
% AIAA Journal, 48(5):865-877, May 2010. doi:10.2514/1.39957

function gopal2010

fd = header('gopal2010_m.set');

% Ria's mail 10-14-2011
% 
% chord = 2cm,
% span (including root cut out) = 4*2 = 8cm
% thickness = 0.05 mm
% U_infinity = 4m/s
% f = 47.75 Hz
% rho_polyester =  1350 kg/m3
%
% Grid motion:
% alpha = pitch_knot*pi/180+pitch_angle*pi/180*cos(omega_t+phase*pi/180)
% flap_angle = flap_amp*pi/180*(cos(omega_t))
% pitch_knot = 12.5 deg
% pitch_angle = 32.5 deg
% phase = 90 deg
% flap_amp = 30 deg
%
% Ria's mail 10-17-2011
%
% I think while making the grid, I have non dimensionalized it wrt c=4cm,
% which is why I have t/c=0.00125.
%
% Sorry about that, could you please use these parameters,
% C=4cm,
% f = 11.94 Hz
% U_infty = 2m/s

E = 2e+9; % N/m^2; polyethilene; 3.5 GPa from the internet
NU = 0.25; % adim; polyethilene, from the internet
RHO = 1350.; % kg/m^3

PRESTRESS = 0.; % N/m

B = 40e-3; % m, chord
L = 3.5*B; % m
H = .05e-3; % m
L0 = 0.5*B; % m

FREQ = 11.94; % Hz
PITCH_0 = 12.5; % deg
PITCH_A = 32.5; % deg
FLAP_A = 30.; % deg
PHASE = 90.; % deg

NB = 10;
% NL = 3.5*NB;
NL = 2*NB;

fprintf(fd, 'set: const integer NL = %d;\n', NL);
fprintf(fd, 'set: const integer NB = %d;\n', NB);
fprintf(fd, '\n');
fprintf(fd, 'set: integer GROUND = 99997;\n'),
fprintf(fd, 'set: integer FLAP = 99998;\n'),
fprintf(fd, 'set: integer PITCH = 99999;\n'),
fprintf(fd, '\n');

fprintf(fd, 'set: const integer NNODES = %d;\n', (NL + 1)*(NB + 1));
fprintf(fd, 'set: const integer NBODIES = %d;\n', (NL + 1)*(NB + 1));
fprintf(fd, 'set: const integer NSHELLS = %d;\n', NL*NB);
fprintf(fd, 'set: const integer NJOINTS = %d;\n', 2*(NL + NB));
fprintf(fd, '\n');

fprintf(fd, 'set: real E = %e; # N/m^2; tune as needed\n', E);
fprintf(fd, 'set: const real NU = %e; # adim\n', NU);
fprintf(fd, 'set: real RHO = %e; # kg/m^3; tune as needed\n', RHO);
fprintf(fd, '\n');

fprintf(fd, 'set: const real L0 = %e; # m\n', L0);
fprintf(fd, 'set: const real L = %e; # m\n', L);
fprintf(fd, 'set: const real B = %e; # m\n', B);
fprintf(fd, 'set: real H = %e; # m; tune as needed\n', H);
fprintf(fd, 'set: real PS = %e; # m\n', PRESTRESS);
fprintf(fd, '\n');

fprintf(fd, 'set: real FREQ = %e*2*pi; # Hz\n', FREQ);
fprintf(fd, 'set: real PITCH_0 = %e*deg2rad; # deg\n', PITCH_0);
fprintf(fd, 'set: real PITCH_A = %e*deg2rad; # deg\n', PITCH_A);
fprintf(fd, 'set: real FLAP_A = %e*deg2rad; # deg\n', FLAP_A);
fprintf(fd, 'set: real PHASE = %e*deg2rad; # deg\n', PHASE);
fprintf(fd, '\n');

footer(fd);

fd = header('gopal2010_m.ref');

fprintf(fd, 'reference: GROUND,\n');
fprintf(fd, '        reference, global, null,\n');
fprintf(fd, '        reference, global, eye,\n');
fprintf(fd, '        reference, global, null,\n');
fprintf(fd, '        reference, global, null;\n');

fprintf(fd, 'reference: FLAP,\n');
fprintf(fd, '        reference, GROUND, null,\n');
fprintf(fd, '        reference, GROUND, eye,\n');
fprintf(fd, '        reference, GROUND, null,\n');
fprintf(fd, '        reference, GROUND, null;\n');

fprintf(fd, 'reference: PITCH,\n');
fprintf(fd, '        reference, FLAP, 0., L0, 0.,\n');
fprintf(fd, '        reference, FLAP, eye,\n');
fprintf(fd, '        reference, FLAP, null,\n');
fprintf(fd, '        reference, FLAP, null;\n');

footer(fd);

fd = header('gopal2010_m.nod');

OFFSET = 100;

for il = 0:NL,
	for ib = 0:NB,
		fprintf(fd, 'structural: %d, dynamic,\n', OFFSET*il + ib);
		fprintf(fd, '        reference, PITCH, %e*B, %e*L, 0.,\n', ib/NB - 0.25, il/NL);
		fprintf(fd, '        reference, PITCH, eye,\n');
		fprintf(fd, '        reference, PITCH, null,\n');
		fprintf(fd, '        reference, PITCH, null;\n');
		fprintf(fd, '\n');
	end
end

footer(fd);

fd = header('gopal2010_m.elm');

fprintf(fd, '# joints\n');
fprintf(fd, '\n');

boundary_nodes = [[1:NL]*OFFSET, [1:NB] + NL*OFFSET, [NL - 1:-1:0]*OFFSET + NB, [NB - 1:-1:0]];

for ib = 1:length(boundary_nodes),
	fprintf(fd, 'joint: %d, total joint,\n', boundary_nodes(ib));
	fprintf(fd, '        PITCH,\n');
	fprintf(fd, '                position, reference, PITCH, null,\n');
	fprintf(fd, '                position orientation, reference, PITCH, eye,\n');
	fprintf(fd, '                rotation orientation, reference, PITCH, eye,\n');
	fprintf(fd, '        %d,\n', boundary_nodes(ib));
	fprintf(fd, '                position, reference, PITCH, null,\n');
	fprintf(fd, '                position orientation, reference, PITCH, eye,\n');
	fprintf(fd, '                rotation orientation, reference, PITCH, eye,\n');
	fprintf(fd, '        position constraint, 1, 1, 1, null,\n');
	fprintf(fd, '        orientation constraint, 1, 1, 1, null;\n');
	fprintf(fd, '\n');
end

fprintf(fd, '# rigid bodies\n');
fprintf(fd, '\n');

for il = 0:NL,
	for ib = 0:NB,
		db = (1/NB);
		dl = (1/NL);
		xb = 0;
		xl = 0;
		if (ib == 0),
			db = db/2;
			xb = ib/NB/4;
		elseif (ib == NB),
			db = db/2;
			xb = -ib/NB/4;
		end
		if (il == 0),
			dl = dl/2;
			xl = il/NL/4;
		elseif (il == NL),
			dl = dl/2;
			xl = -il/NL/4;
		end
		m = db*dl;
		Jx = db^2*m/12;
		Jz = dl^2*m/12;

		fprintf(fd, 'body: %d, %d,\n', OFFSET*il + ib, OFFSET*il + ib);
		fprintf(fd, '        %e*B*L*H*RHO,\n', m);
		fprintf(fd, '        reference, node, %e*B, %e*L, 0.,\n', xb, xl);
		fprintf(fd, '        diag, %e*B*L^3*H*RHO, %e*B^3*L*H*RHO, (%e*B^3*L + %e*B*L^3)*H*RHO;\n', Jz, Jx, Jz, Jx);
		fprintf(fd, '\n');
	end
end

fprintf(fd, 'inertia: 0, body, all;\n');
fprintf(fd, '\n');

fprintf(fd, '# shells\n');
fprintf(fd, '\n');

for il = 1:NL,
	for ib = 1:NB,
		label = OFFSET*il + ib;
		node1 = OFFSET*il + ib;
		node2 = OFFSET*(il - 1) + ib;
		node3 = OFFSET*(il - 1) + (ib - 1);
		node4 = OFFSET*il + (ib - 1);
		fprintf(fd, 'shell4easans: %d, %d, %d, %d, %d,\n', ...
			label, node1, node2, node3, node4);
		fprintf(fd, '        isotropic, E, E, nu, NU, thickness, H,\n');
		fprintf(fd, '        prestress, PS, 0., 0., 0., PS, 0., 0., 0., 0., 0., 0., 0.;\n');
	end
end

footer(fd);

fd = header('gopal2010_ext_m.elm');

mb_file = 'gopal2010_shell_mb.dat';
cfd_file = 'gopal2010_shell_surface_grid.dat';
mapping_file = 'gopal2010_shell_H.dat';

fprintf(fd, 'force: 111, external structural mapping,\n');
fprintf(fd, '        socket,\n');
fprintf(fd, '        create, yes,\n');
fprintf(fd, '        path, "$MBSOCK",\n');
fprintf(fd, '        no signal,\n');
fprintf(fd, '        coupling, tight,\n');
fprintf(fd, '        reference node, PITCH,\n');
fprintf(fd, '        orientation, orientation matrix,\n');
fprintf(fd, '        use reference node forces, no,\n');
fprintf(fd, '        points number, 2*NNODES,\n');
for il = 0:NL,
	for ib = 0:NB,
		fprintf(fd, '                %4d, offset, 0., 0., .25*H, offset, 0., 0., -.25*H,\n', OFFSET*il + ib);
	end
end
fprintf(fd, '        # echo, "%s", precision, 16, surface, "%s", output, "%s", order, 2, basenode, 12, weight, 2, stop;\n', mb_file, cfd_file, mapping_file);
fprintf(fd, '        mapped points number, from file,\n');
fprintf(fd, '        sparse mapping file, "%s";\n', mapping_file);

footer(fd);

endfunction



function fd = header(fname);

fd = fopen(fname, 'w');

fprintf(fd, '# MBDyn (C) is a multibody analysis code. \n');
fprintf(fd, '# http://www.mbdyn.org\n');
fprintf(fd, '# \n');
fprintf(fd, '# Copyright (C) 1996-2017\n');
fprintf(fd, '# \n');
fprintf(fd, '# Pierangelo Masarati	<masarati@aero.polimi.it>\n');
fprintf(fd, '# Paolo Mantegazza	<mantegazza@aero.polimi.it>\n');
fprintf(fd, '# \n');
fprintf(fd, '# Dipartimento di Ingegneria Aerospaziale - Politecnico di Milano\n');
fprintf(fd, '# via La Masa, 34 - 20156 Milano, Italy\n');
fprintf(fd, '# http://www.aero.polimi.it\n');
fprintf(fd, '# \n');
fprintf(fd, '# Changing this copyright notice is forbidden.\n');
fprintf(fd, '# \n');
fprintf(fd, '# This program is free software; you can redistribute it and/or modify\n');
fprintf(fd, '# it under the terms of the GNU General Public License as published by\n');
fprintf(fd, '# the Free Software Foundation (version 2 of the License).\n');
fprintf(fd, '# \n');
fprintf(fd, '# \n');
fprintf(fd, '# This program is distributed in the hope that it will be useful,\n');
fprintf(fd, '# but WITHOUT ANY WARRANTY; without even the implied warranty of\n');
fprintf(fd, '# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n');
fprintf(fd, '# GNU General Public License for more details.\n');
fprintf(fd, '# \n');
fprintf(fd, '# You should have received a copy of the GNU General Public License\n');
fprintf(fd, '# along with this program; if not, write to the Free Software\n');
fprintf(fd, '# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA\n');
fprintf(fd, '#\n');

endfunction



function footer(fd)

fprintf(fd, '\n');
fprintf(fd, '# vim%s\n', ':ft=mbd');

fclose(fd);

endfunction

File: tests/plates/gopal2010/gopal2010.py
#! /usr/bin/python
#
# $Header: /var/cvs/mbdyn/mbdyn/mbdyn-1.0/tests/plates/gopal2010/gopal2010.py,v 1.2 2014/12/13 17:15:55 masarati Exp $
#
# MBDyn (C) is a multibody analysis code. 
# http://www.mbdyn.org
# 
# Copyright (C) 1996-2010
# 
# Pierangelo Masarati	<masarati@aero.polimi.it>
# Paolo Mantegazza	<mantegazza@aero.polimi.it>
# 
# Dipartimento di Ingegneria Aerospaziale - Politecnico di Milano
# via La Masa, 34 - 20156 Milano, Italy
# http://www.aero.polimi.it
# 
# Changing this copyright notice is forbidden.
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation (version 2 of the License).
# 
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
# 

import sys;
import os;
import tempfile;

sys.path.append(os.path.abspath('/usr/local/mbdyn/libexec/mbpy/'));

from mbc_py_interface import mbcNodal
from numpy import *

# note: need mbc_py_interface.py, mbc_py.py and _mbc_py.so in PYTHONPATH
# right now they are all installed in $(MBDYN_libexecdir)/mbpy
# execute after setting PYTHONPATH=$(MBDYN_libexecdir)/mbpy

tmpdir = tempfile.mkdtemp('', '.mbdyn_');
path = tmpdir + '/mbdyn.sock';
os.environ['MBSOCK'] = path;

#inputfile = 'gopal2010_ext';
inputfile = 'gopal2010m_ext';

mbdyn_cmd = 'mbdyn';
os.system(mbdyn_cmd + ' -f ' + inputfile + ' -o output > output.txt 2>&1 &');

host = "";
port = 0;
timeout = -1;
verbose = 0;
data_and_next = 1;
refnode = 1;
nodes = 34595 # set to the actual count of mapped points
labels = 0;
# rot = 0x1000;	# reference node orientation vector; no node orientations
rot = 0x2000;	# reference node orientation matrix; no node orientations
# rot = 0x4000;	# reference node Euler 123; no node orientations
accels = 0;
structure = mbcNodal(path, host, port, timeout, verbose, data_and_next, refnode, nodes, labels, rot, accels);

# note: could fail!
structure.negotiate();

# structure.r_*: reference node stuff (if configured)
# structure.n_*: node stuff

# input:
# structure.r_x: reference node position
# structure.r_theta: reference node orientation vector (if rot == 0x1000)
# structure.r_r: reference node orientation matrix (if rot == 0x2000)
# structure.r_euler_123: reference node Euler angles 123 (if rot == 0x4000)
# structure.r_xp: reference node velocity
# structure.r_omega: reference node angular velocity
# structure.r_xpp: reference node acceleration (if accels != 0)
# structure.r_omegap: reference node angular acceleration (if accels != 0)

# output:
# structure.r_f: reference node force
# structure.r_m: reference node moment

# input:
# structure.n_x: nodes position
# structure.n_xp: nodes velocity
# structure.n_xpp: nodes acceleration (if accels != 0)

# output:
# structure.n_f: nodes force

# examples:
# position of reference node: structure.r_x[0], structure.r_x[1], structure.r_x[2]
# position of node 7: structure.n_x[7*3-3], structure.n_x[7*3-2], structure.n_x[7*3-1]

# more than one communicator can be used in the same analysis

structure.r_f[0] = 0.;
structure.r_f[1] = 0.;
structure.r_f[2] = 0.;
structure.r_m[0] = 0.;
structure.r_m[1] = 0.;
structure.r_m[2] = 0.;

#xmprev = zeros((3*nodes, 1));
#xmcurr = zeros((3*nodes, 1));
#xm = zeros((3*nodes, 1));

# damp = -2e-8;
damp = -2e-7;
# damp = -2e-6;
# damp = -2e-5;
cnt = 0;
iter_cnt = 0;
while 1:
	if (structure.recv()):
		break;

	x0 = reshape(structure.r_x[0:3], (3, 1));
	R0 = reshape(structure.r_r, (3, 3), 'F');
	x0p = dot(R0.transpose(), structure.r_xp);
	omega0 = dot(R0.transpose(), structure.r_omega);

	#for n in range(nodes):
	#	xmcurr[3*n:3*(n+1)] = dot(R0, reshape(structure.n_x[3*n:3*(n+1)], (3, 1)));
	#if (cnt > 0):
	#	xm = reshape(xmcurr, (3*nodes, 1)) - reshape(xmprev, (3*nodes, 1));
	#xmprev = xmcurr.copy();

	print '#'
	#xm = ravel(xm);
	#for n in range(nodes):
	#	print xm[3*n], xm[3*n+1], xm[3*n+2]

	for n in range(nodes):
		structure.n_f[3*n:3*(n+1)] = damp*(x0p + cross(omega0, structure.n_x[3*n:3*(n+1)]) + structure.n_xp[3*n:3*(n+1)]);

	if iter_cnt == 5:
		print '# cnt', cnt
		for n in range(nodes):
			x = ravel(x0 + dot(R0, reshape(structure.n_x[3*n:3*(n+1)], (3, 1))));
			print x[0], x[1], x[2]
			# print structure.n_x[3*n], structure.n_x[3*n + 1], structure.n_x[3*n + 2];
	
	# send(1) at convergence; send(0) for another iteration
	if (structure.send(iter_cnt == 5)):
		break;

	if iter_cnt == 5:
		iter_cnt = 0;
		cnt = cnt + 1;

	iter_cnt = iter_cnt + 1;

structure.destroy();
os.rmdir(tmpdir);

